include/rmm/cuda_event.hpp

```diff
@@ -0,0 +1,257 @@
+   * the owned event.
+   */
+  cuda_event(cuda_event&&) = default;
+  /**
+   * @brief Move copy assignment operator (default)
+   *
+   * A moved-from cuda_event is invalid and it is Undefined Behavior to call methods that access
+   * the owned event.
+   */
+  cuda_event& operator=(cuda_event&&) = default;
+  ~cuda_event()                       = default;
+  cuda_event(cuda_event const&)       = delete;  // Copying disallowed: one event one owner
+  cuda_event& operator=(cuda_event&) = delete;
+
+  /**
+   * @brief Construct a new cuda event object
+   *
+   * @throw rmm::cuda_error if event creation fails
+   */
+  cuda_event()
+    : event_{[]() {
+               cudaEvent_t* s = new cudaEvent_t;
+               RMM_CUDA_TRY(cudaEventCreate(s));
+               return s;
+             }(),
+             [](cudaEvent_t* s) {
+               RMM_ASSERT_CUDA_SUCCESS(cudaEventDestroy(*s));
+               delete s;
+             }}
+  {
+  }
+
+  /**
+   * @brief Construct a new cuda event object
+   *
+   * @throw rmm::cuda_error if event creation fails
+   */
+  cuda_event(unsigned int flags)
+    : event_{[flags]() {
+               cudaEvent_t* s = new cudaEvent_t;
+               RMM_CUDA_TRY(cudaEventCreateWithFlags(s, flags));
+               return s;
+             }(),
+             [](cudaEvent_t* s) {
+               RMM_ASSERT_CUDA_SUCCESS(cudaEventDestroy(*s));
+               delete s;
+             }}
+  {
+  }
+
+  /**
+   * @brief Returns true if the owned event is non-null
+   *
+   * @return true If the owned event has not been explicitly moved and is therefore non-null.
+   * @return false If the owned event has been explicitly moved and is therefore null.
+   */
+  bool is_valid() const { return event_ != nullptr; }
+
+  /**
+   * @brief Get the value of the wrapped CUDA event.
+   *
+   * @return cudaEvent_t The wrapped CUDA event.
+   */
+  cudaEvent_t value() const
+  {
+    RMM_LOGGING_ASSERT(is_valid());
+    return *event_;
+  }
+
+  /**
+   * @brief Explicit conversion to cudaEvent_t.
+   */
+  explicit operator cudaEvent_t() const noexcept { return value(); }
+
+  /**
+   * @brief Record the owned CUDA event in the given CUDA stream.
+   *
+   * Calls `cudaEventRecord()`.
+   *
+   * @throw rmm::cuda_error if event recording fails
+   */
+  void record(cudaStream_t stream) const { RMM_CUDA_TRY(cudaEventRecord(value(), stream)); }
+
+  /**
+   * @brief Record the owned CUDA event in the given CUDA stream.
+   *
+   * Calls `cudaEventRecord()`.
+   *
+   * @throw rmm::cuda_error if event recording fails
+   */
+  void record(cuda_stream_view stream) const { record(stream.value()); }
+
+  /**
+   * @brief Record the owned CUDA event in the given CUDA stream.
+   *
+   * Calls `cudaEventRecordWithFlags()`.
+   *
+   * @throw rmm::cuda_error if event recording fails
+   */
+  void record(cudaStream_t stream, unsigned int flags) const
```

Comment:
We shouldn't expose the raw `unsigned int` flags. We should instead have some kind of appropriately named `enum` and map that to the raw flags.